### PR TITLE
try-runtime-cli: improve ci stability 

### PR DESCRIPTION
**Pull request by liamaharon**
- Updates `HttpClientBuilder` config to fix some stability issues with the CI and increase the max possible batch size
-  Limits parallelism to 4 threads. Performance improvement beyond 4 is negligible, and can lead to some weird stability issues with the `HttpClient` (https://github.com/hyperium/hyper/issues/2112)
    - Ideally we would refactor the code to remove all threading and rely on tokio futures for request parallelism (or use just two threads: one for inserting keys and one dedicated for http requests using tokio futures). However limiting the number of threads seems sufficient for fixing the CI issue for now. Larger refactor will come later, probably when we implement lazy-download.
- CI tested against this branch here: https://github.com/paritytech/polkadot/pull/7133 
  - westend still failing, *not* due to download but due to an on-runtime-upgrade try-runtime hook actually failing: `the on_chain version is equal or more than the current one`. I guess we need to update its Migrations struct? If so, I'm happy to handle that in a seperate PR.

Closes https://github.com/paritytech/substrate/issues/14013 and https://github.com/paritytech/substrate/issues/14006

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2741749

**Review by niklasad1 — APPROVED** [review]
Looks good to me

It would be safer to create a new HttpClient in each thread spawned in the "remote externalities" than to limit the number of threads, fine for now

**Comment by liamaharon** [comment]
> Looks good to me
> 
> It would be safer to create a new HttpClient in each thread spawned in the "remote externalities" than to limit the number of threads, fine for now

Yup I was considering this but it seems like it would require a larger refactor. We'll be doing a large refactor later anyway for lazy-download, and CI seems happy with this configuration, so decided not to do it here.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2742278

File: utils/frame/remote-externalities/src/lib.rs [hunk 1]
```diff
@@ -157,10 +157,14 @@ impl Transport {
 			} else {
 				uri.clone()
 			};
-			let http_client = HttpClientBuilder::default().build(uri).map_err(|e| {
-				log::error!(target: LOG_TARGET, "error: {:?}", e);
-				"failed to build http client"
-			})?;
+			let http_client = HttpClientBuilder::default()
+				.max_request_body_size(u32::MAX)
+				.request_timeout(std::time::Duration::from_secs(60 * 5))
```
Comment (ggwpez) [inline utils/frame/remote-externalities/src/lib.rs:162]: Is this needed? I would rather have it fail after 1 min if the node is down, and not wait another 4.

File: utils/frame/remote-externalities/src/lib.rs [hunk 2]
```diff
@@ -323,16 +327,23 @@ where
...
 	/// Get the number of threads to use.
+	/// Cap the number of threads. Performance improvement beyond a small number of threads is
+	/// negligible, and too many threads can create issues with the HttpClient.
 	fn threads() -> NonZeroUsize {
-		thread::available_parallelism()
-			.unwrap_or(NonZeroUsize::new(1usize).expect("4 is non-zero; qed"))
+		let avaliable = thread::available_parallelism()
+			.unwrap_or(NonZeroUsize::new(1usize).expect("1 is non-zero; qed"))
+			.get();
+		assert!(avaliable > 0, "avaliable parallelism must be greater than 0");
```
Comment (ggwpez) [inline utils/frame/remote-externalities/src/lib.rs:344]: You could also add an env var check like `TRY_RUNTIME_MAX_THREADS`, that would be easier than to add a CLI argument, but a bit hacky.

**Review by ggwpez — APPROVED** [review]
Thanks!

File: utils/frame/remote-externalities/src/lib.rs (hunk 1 above)
Comment (niklasad1) [reply utils/frame/remote-externalities/src/lib.rs:162]: You can remove if you want but if you do big batch request is could take longer :P

We bumped into this in the staking miner against public RPC nodes IIRC

File: utils/frame/remote-externalities/src/lib.rs (hunk 2 above)
Comment (liamaharon) [reply utils/frame/remote-externalities/src/lib.rs:344]: Sure, added a `TRY_RUNTIME_MAX_THREADS` env var

File: utils/frame/remote-externalities/src/lib.rs (hunk 1 above)
Comment (ggwpez) [reply utils/frame/remote-externalities/src/lib.rs:162]: Yea but we have the re-try feature :sunglasses:

**Pull request merged**